base_layer/core/tests/tests/mempool.rs

```diff
@@ -330,6 +334,135 @@ async fn test_time_locked() {
+    let height = header.height;
+
+    let mut transactions_new = Vec::with_capacity(transactions.len());
+    for txn in transactions {
+        transactions_new.push(Transaction {
+            offset: txn.offset,
+            body: {
+                let mut inputs = Vec::with_capacity(txn.body.inputs().len());
+                for input in txn.body.inputs().iter() {
+                    inputs.push(input.clone());
+                }
+                let mut outputs = Vec::with_capacity(txn.body.outputs().len());
+                for output in txn.body.outputs().iter() {
+                    outputs.push(output.clone());
+                }
+                let mut kernels = Vec::with_capacity(txn.body.kernels().len());
+                for kernel in txn.body.kernels().iter() {
+                    kernels.push(TransactionKernel {
+                        version: kernel.version,
+                        features: kernel.features,
+                        fee: (u64::MAX / 2).into(), // This is the adversary's attack!
+                        lock_height: kernel.lock_height,
+                        excess_sig: kernel.excess_sig.clone(),
+                        excess: kernel.excess.clone(),
+                        burn_commitment: kernel.burn_commitment.clone(),
+                    });
+                }
+                AggregateBody::new(inputs, outputs, kernels)
+            },
+            script_offset: txn.script_offset,
+        });
+    }
+    let template = NewBlockTemplate::from_block(
+        header
+            .into_builder()
+            .with_transactions(transactions_new)
+            .with_coinbase_utxo(coinbase_utxo, coinbase_kernel)
+            .build(),
+        Difficulty::min(),
+        consensus_manager.get_block_reward_at(height),
+    );
+    let new_block = store.prepare_new_block(template).unwrap();
+
+    // This will call `BlockBuilder::add_kernels(...)` and `AggregateBody::get_total_fee(...)`, which will overflow if
+    // regressed
+    let block_result = store.add_block(new_block.into());
+    assert!(block_result.is_err());
+    assert_eq!(
+        block_result.unwrap_err().to_string(),
+        "Validation error: The total value + fees of the block exceeds the maximum allowance on chain".to_string()
+    );
+
+    let schema = txn_schema!(
+        from: vec![outputs[1][3].clone()],
+        to: vec![],
+        fee: MicroMinotari(u64::MAX / 2), // This is the adversary's attack!
+        lock: 0,
+        features: OutputFeatures::default()
+    );
+    let stx_builder = create_stx_protocol_internal(schema, &key_manager, &mut Vec::new()).await;
+
+    // This will call `Fee::calculate(...)`, which will overflow if regressed
+    let build_result = stx_builder.build().await;
+    assert!(build_result.is_err());
+    assert_eq!(
+        format!("{:?}", build_result.unwrap_err()),
+        "You are spending (18446744073709.550781 T) more than you're providing (10.000000 T)."
```

Comment:
This does not seem correct. How did it get to this amount?
u64 max = "18446744073709.551615"
So this test is seems if its not running into overflow

this should print out 18446744073709.551615 - 10 = 18446744073691.551615

Reply:
The error message is misleading:
```rust
        let total_input_value = [total_to_self, total_amount, fee_without_change]
            .iter()
            .fold(Ok(MicroMinotari::zero()), |acc, x| {
                acc?.checked_add(x).ok_or("Total input value overflow")
            })?;
        let change_amount = total_being_spent.checked_sub(total_input_value);
        match change_amount {
            None => Err(format!(
                "You are spending ({}) more than you're providing ({}).",
                total_input_value, total_being_spent
            )),
```

Reply:
This is the part that will overflow if unbounded, :
```rust
        let fee_without_change = self.fee().calculate(
            fee_per_gram,
            1,
            num_inputs,
            num_outputs,
            features_and_scripts_size_without_change,
        );
```

Reply:
For this test the inputs are
```
total_to_self 0 µT, total_amount 0 µT, fee_without_change 18446744073709.550781 T
```
resulting in  error message
```
"You are spending more than you're providing: provided 10.000000 T, required 18446744073709.550781 T."
```

Reply:
Ok, found the error with the displayed value being off from `u64::MAX`; the is was in `impl Display for MicroMinotari` as it did floating point arithmetic. I changed it to rather do string manipulation as we only need to insert a decimal point in the string representation.

Reply:
Now the test values are:
```
total_to_self 0 µT, total_amount 0 µT, fee_without_change 18446744073709.551615 T
```
resulting in error message
```
"You are spending more than you're providing: provided 10.000000 T, required 18446744073709.551615 T."
```
which is consistent with `u64::MAX`

Reply:
Good find, that far not ideal, especially when it's displaying amounts, we should remove that rounding. But not for this PR.
Opened a new issue: https://github.com/tari-project/tari/issues/5839